node/src/components/consensus/era_supervisor/tests.rs

```diff
@@ -0,0 +1,120 @@
+    testing::TestRng,
+    utils::{self, External, Loadable},
+};
+
+type ClMessage = Message<ClContext>;
+
+#[derive(Debug, From)]
+enum Event {
+    #[from]
+    Consensus(super::Event<NodeId>),
+    #[from]
+    Network(NetworkRequest<NodeId, protocol::Message>),
+    #[from]
+    BlockProposer(BlockProposerRequest),
+    #[from]
+    ConsensusAnnouncement(ConsensusAnnouncement),
+    #[from]
+    BlockExecutor(BlockExecutorRequest),
+    #[from]
+    BlockValidator(BlockValidationRequest<ProtoBlock, NodeId>),
+    #[from]
+    Storage(StorageRequest<Storage>),
+    #[from]
+    ContractRuntime(ContractRuntimeRequest),
+}
+
+/// Loads the local chainspec and overrides timestamp and genesis account with the given stakes.
+fn new_test_chainspec(stakes: Vec<(PublicKey, u64)>) -> Chainspec {
+    let mut chainspec = Chainspec::from_resources("local/chainspec.toml");
+    chainspec.genesis.accounts = stakes
+        .into_iter()
+        .map(|(pk, stake)| {
+            let motes = Motes::new(stake.into());
+            GenesisAccount::new(pk.into(), pk.to_account_hash(), motes, motes)
+        })
+        .collect();
+    chainspec.genesis.timestamp = Timestamp::now() + 1000.into();
+    chainspec.genesis.highway_config.genesis_era_start_timestamp = chainspec.genesis.timestamp;
+    chainspec
+}
+
+#[test]
+fn era_supervisor() -> Result<(), Error> {
+    let mut rng = TestRng::new();
+
+    let alice_sk = SecretKey::random(&mut rng);
+    let alice_pk = PublicKey::from(&alice_sk);
+    let bob_sk = SecretKey::random(&mut rng);
+    let bob_pk = PublicKey::from(&bob_sk);
+
+    let chainspec = new_test_chainspec(vec![(alice_pk, 100)]);
+    let config = Config {
+        secret_key_path: External::Loaded(alice_sk),
+    };
+
+    let registry = Registry::new();
+    let scheduler = utils::leak(Scheduler::<Event>::new(QueueKind::weights()));
+    let event_queue = EventQueueHandle::new(&scheduler);
+    let effect_builder = EffectBuilder::new(event_queue);
+
+    let (mut es, effects) = EraSupervisor::new(
+        Timestamp::now(),
+        WithDir::new("tmp", config),
+        effect_builder,
+        vec![(alice_pk, Motes::new(100.into()))],
+        &chainspec,
+        Default::default(), // genesis state root hash
+        &registry,
+        Box::new(MockProto::new_boxed),
+        &mut rng,
+    )?;
+    assert!(effects.is_empty());
+
+    let mut handle = |es: &mut EraSupervisor<NodeId>, event: super::Event<NodeId>| {
+        es.handle_event(effect_builder, &mut rng, event)
+    };
+
+    let event = ClMessage::BlockByOtherValidator {
+        value: CandidateBlock::new(ProtoBlock::new(vec![], true), vec![alice_pk]),
+        timestamp: Timestamp::now(),
+        proposer: bob_pk,
+    }
+    .received(NodeId(1), EraId(0));
+    handle(&mut es, event);
+    let event = ClMessage::FinalizeBlock.received(NodeId(1), EraId(0));
+    let effects = handle(&mut es, event);
+    assert_eq!(effects.len(), 2);
```

Comment:
I think my n00b card is expiring soon, but what are the effects we can expect here?  Would pattern matching on them be a little brittle, since we might expect the effects to change a bit with churn?

Reply:
Yes, I was meaning to do that yesterday, but effects are futures, so it's a bit complicated. I plan to do it as part of [NDRS-512](https://casperlabs.atlassian.net/browse/NDRS-512).